File: R/1_generate_synthetic.r
source("R/0_setup.r")
library(data.table)
library(magrittr)
# library(compiler)
# enableJIT(3)

#require(data.table)
#' Generate Synthetic dat
#' @import data.table
gen_datatable_synthetic <- function(N=2e9/8, K=100) {
  data.table(
    id1 = sample(sprintf("id%03d",1:K), N, TRUE),      # large groups (char)
    id2 = sample(sprintf("id%03d",1:K), N, TRUE),      # large groups (char)
    id3 = sample(sprintf("id%010d",1:(N/K)), N, TRUE), # small groups (char)
    id4 = sample(K, N, TRUE),                          # large groups (int)
    id5 = sample(K, N, TRUE),                          # large groups (int)
    id6 = sample(N/K, N, TRUE),                        # small groups (int)
    v1 =  sample(5, N, TRUE),                          # int in range [1,5]
    v2 =  sample(5, N, TRUE),                          # int in range [1,5]
    v3 =  sample(round(runif(100,max=100),4), N, TRUE) # numeric e.g. 23.5749
  )
}


r_syn_gen_test <- function() {
  res <- NULL
  DT <- NULL
  res <- c(res, list(system.time(DT <- gen_datatable_synthetic())))
  cat("GB =", round(sum(gc()[,2])/1024, 3), "\n")
  res <- c(res, list(system.time( DT[, sum(v1), keyby=id1] )))
  res <- c(res, list(system.time( DT[, sum(v1), keyby=id1] )))
  res <- c(res, list(system.time( DT[, sum(v1), keyby="id1,id2"] )))
  res <- c(res, list(system.time( DT[, sum(v1), keyby="id1,id2"] )))
  res <- c(res, list(system.time( DT[, list(sum(v1),mean(v3)), keyby=id3] )))
  res <- c(res, list(system.time( DT[, list(sum(v1),mean(v3)), keyby=id3] )))
  res <- c(res, list(system.time( DT[, lapply(.SD, mean), keyby=id4, .SDcols=7:9] )))
  res <- c(res, list(system.time( DT[, lapply(.SD, mean), keyby=id4, .SDcols=7:9] )))
  res <- c(res, list(system.time( DT[, lapply(.SD, sum), keyby=id6, .SDcols=7:9] )))
  res <- c(res, list(system.time( DT[, lapply(.SD, sum), keyby=id6, .SDcols=7:9] )))
  
  
  res1 <- lapply(res, function(rr) rr %>% lapply(function(x) x) %>% as.data.table) %>% rbindlist
  setDT(res1)
  #browser()
  #res1[,test := c("generate_data", paste0("test",1:11))]
  print(res1)
  res1
}

#' Run R synthetic data generation test n times
#' @export
# n_r_syn_gen_test <- function(n) {
#   res2 <- lapply(1:n, function(i) {
#     print(i)
#     r_syn_gen_test()
#   })
#   return(res2)
# }
# 
# system.time(res <- n_r_syn_gen_test(1))

res100 <- NULL
for(i in 1:100) {
  res100 <- c(res100, list(r_syn_gen_test()))
}

res101 <- lapply(res100, function(x) {
  y = copy(x)
  setDT(y)
  y[,test:= paste0("test",1:.N)]
  y
}) %>% rbindlist
write.csv(res101,file.path("Results",paste0("r ",gsub(":","",Sys.time()),".csv")))

# 
# 
# res101[,.N]
# 
# res101[test %in% c("test2", "test3"), mean(elapsed)]

File: Julia/1_generate_synthetic.jl
# Sashi' IndexedTable version
using Distributions, PooledArrays, Nulls
#using DataFrames

#using IndexedTables

const N=Int64(2e9); const K=100;

const pool = [@sprintf "id%03d" k for k in 1:K]
const pool1 = [@sprintf "id%010d" k for k in 1:(N/K)]
#const pool_withmissing = vcat(null, pool)
#const pool1_withmssing = vcat(null, pool1)

function randstrarray(pool, N)
    PooledArray(PooledArrays.RefArray(rand(UInt8(1):UInt8(K), N)), pool)
end

using JuliaDB
srand(1)

DT = IndexedTable(
  Columns(
    row_id = [1:N;]
    ),
  Columns(
    id1 = randstrarray(pool, N),
    id2 = randstrarray(pool, N),
    id3 = randstrarray(pool1, N),
    id4 = rand(1:K, N),                          # large groups (int)
    id5 = rand(1:K, N),                          # large groups (int)
    id6 = rand(1:(N/K), N),                        # small groups (int)
    v1 =  rand(1:5, N),                          # int in range [1,5]
    v2 =  rand(1:5, N),                          # int in range [1,5]
    v3 =  rand(round.(rand(Uniform(0,100),100),4), N) # numeric e.g. 23.5749
    ));

# using DataFrames
# srand(1)
# @time dtdf = DataFrame(
#   :id1 => randstrarray(pool, N),
#   :id2 => randstrarray(pool, N),
#   :id3 => randstrarray(pool1, N),
#   :id4 => rand(1:K, N),                          # large groups (int)
#   :id5 => rand(1:K, N),                          # large groups (int)
#   :id6 => rand(1:(N/K), N),                        # small groups (int)
#   :v1 =>  rand(1:5, N),                          # int in range [1,5]
#   :v2 =>  rand(1:5, N),                          # int in range [1,5]
#   :v3 =>  rand(round.(rand(Uniform(0,100),100),4), N) # numeric e.g. 23.5749
#  );
# srand(1)
# @time dtdf = DataFrame(
#   id1 = randstrarray(pool, N),
#   id2 = randstrarray(pool, N),
#   id3 = randstrarray(pool1, N),
#   id4 = rand(1:K, N),                          # large groups (int)
#   id5 = rand(1:K, N),                          # large groups (int)
#   id6 = rand(1:(N/K), N),                        # small groups (int)
#   v1 =  rand(1:5, N),                          # int in range [1,5]
#   v2 =  rand(1:5, N),                          # int in range [1,5]
#   v3 =  rand(round.(rand(Uniform(0,100),100),4), N) # numeric e.g. 23.5749
# );


# Oringal version
# using DataFrames, Distributions, BenchmarkTools;
#
# function gen_dict(N,K)
#     Dict(
#       :id1 => sample([@sprintf "id%03d" k for k in 1:K], N),      # large groups (char)
#       :id2 => sample([@sprintf "id%03d" k for k in 1:K], N),      # large groups (char)
#       :id3 => sample([@sprintf "id%010d" k for k in 1:(N/K)], N), # small groups (char)
#       :id4 => sample(1:K, N),                          # large groups (int)
#       :id5 => sample(1:K, N),                          # large groups (int)
#       :id6 => sample(1:(N/K), N),                        # small groups (int)
#       :v1 =>  sample(1:5, N),                          # int in range [1,5]
#       :v2 =>  sample(1:5, N),                          # int in range [1,5]
#       :v3 =>  sample(round.(rand(Uniform(0,100),100),4), N) # numeric e.g. 23.5749
#     )
# end
#
# function gen_datatable_synthetic(N, K)
#     DataFrame(;gen_dict(N,K)...)
# end
#
# # @benchmark d1 = gen_dict(1,1)
# # @benchmark d1 = gen_dict(N,K)
# @time DT_orig = gen_datatable_synthetic(N,K);
# @benchmark DT_orig = gen_datatable_synthetic(N,K);
